ARMC6: Add a build profile extension with the link-time optimizer enabled

fkjagodzinski opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->
### Description (*required*)

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
##### Summary of change (*What the change is for and why*)
Added an `lto` build profile extension for `ARMC6` toolchain.
* added `-flto` to `common` flags,
* added `--lto` & `--lto_level=Oz` to `ld` flags.

###### Results
`Product: ARM Compiler 6.11 Professional Component: ARM Compiler 6.11 Tool: armclang [5d3b4200]`

#### mbed-os-example-blinky@mbed-os-5.14.1
Mbed OS branch|commit SHA|RAM (data + bss)|Flash (text + data)|build time|details
---|---|---:|---:|---:|---
*mbed-os-5.14.1*|`679d248`|206047|39168|739.53|[blinky-679d248.txt](https://github.com/ARMmbed/mbed-os/files/3851849/blinky-679d248.txt)
*armc6_build-enable_lto_for_release*|`c6a655f`|205848|36601|610.81|[blinky-c6a655f.txt](https://github.com/ARMmbed/mbed-os/files/3851850/blinky-c6a655f.txt)

#### mbed-cloud-client-example@4.0.0
Mbed OS branch|commit SHA|RAM (data + bss)|Flash (text + data)|build time|details
---|---|---:|---:|---:|---
*mbed-os-5.14.1*|`679d248`|252457|361038|828.68|[ccexample-679d248.txt](https://github.com/ARMmbed/mbed-os/files/3851851/ccexample-679d248.txt)
*armc6_build-enable_lto_for_release*|`c6a655f`|251580|329505|835.21|[ccexample-c6a655f.txt](https://github.com/ARMmbed/mbed-os/files/3851873/ccexample-c6a655f.txt)

###### Build commands used to produce the above results
*mbed-os-example-blinky*
```bash
cd $(mktemp -d)
git clone git@github.com:ARMmbed/mbed-os-example-blinky.git
cd mbed-os-example-blinky/
git checkout mbed-os-5.14.1
git clean -dxff && git reset --hard

# Define Mbed OS rev
# 1. mbed-os-5.14.1
echo 'https://github.com/ARMmbed/mbed-os/#679d24833acf0a0b5b0d528576bb37c70863bc4e' > mbed-os.lib
# 2. LTO branch on top of mbed-os-5.14.1
echo 'https://github.com/fkjagodzinski/mbed-os/#d7df3e673144232d3e162de19ec13f0472d725c6' > mbed-os.lib

mbed deploy
cd mbed-os/ && git log --oneline -1 && cd ..
time -p mbed compile -t ARM -m K64F --profile mbed-os/tools/profiles/release.json

# More details
mbed compile -t ARM -m K64F --profile mbed-os/tools/profiles/release.json --stats-depth=100
```
*mbed-cloud-client-example*
```bash
cd $(mktemp -d)
git clone git@github.com:ARMmbed/mbed-cloud-client-example.git
cd mbed-cloud-client-example/
git checkout 4.0.0 
git clean -dxff && git reset --hard
cp ../mbed_cloud_dev_credentials.c .

# Define Mbed OS rev
# 1. mbed-os-5.14.1
echo 'https://github.com/ARMmbed/mbed-os/#679d24833acf0a0b5b0d528576bb37c70863bc4e' > mbed-os.lib
# 2. LTO branch on top of mbed-os-5.14.1
echo 'https://github.com/fkjagodzinski/mbed-os/#d7df3e673144232d3e162de19ec13f0472d725c6' > mbed-os.lib

mbed deploy
cd mbed-os/ && git log --oneline -1 && cd ..
time -p mbed compile -t ARM -m K64F --profile mbed-os/tools/profiles/release.json

# More details
mbed compile -t ARM -m K64F --profile mbed-os/tools/profiles/release.json --stats-depth=100
```

##### Documentation (*Details of any document updates required*)

----------------------------------------------------------------------------------------------------------------
### Pull request type (*required*)

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results (*required*)

<!--
    Required
    For example, add test results for new target
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers (*optional*)

<!--
    Optional
    Request additional reviewers with @username
-->
@jamesbeyond @kjbracey-arm @maciejbocianski 

----------------------------------------------------------------------------------------------------------------
### Release Notes (*required for feature/major PRs*)

<!--
    All 3 sections are compulsory for Major PR types. For Feature PRs only the summary section is required.
    This section is automatically added to release notes. Please fill in each sub-section with sufficient detail for a user.
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types). 
-->

##### Summary of changes
Add a build profile extension with the link-time optimizer enabled for ARMC6 toolchain.

##### Impact of changes

##### Migration actions required

fkjagodzinski committed:
ARMC6: Enable link-time optimizer for release profile

bulislaw reviewed (commented):
Should we also do it for develop?

fkjagodzinski committed:
ARMC6: Enable link-time optimizer for develop profile

fkjagodzinski committed:
NUVOTON: Fix undefined reference to Reset_Handler_Cascade

Add a "used" attribute to Reset_Handler_Cascade to fix GCC build with
the "-flto" flag.
This attribute, attached to a function, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
STM: change rtc irq handler name

Fix for the error caused by lto on armc6 compiler:

L6137E: Symbol RTC_IRQHandler was not preserved by the LTO codegen but is needed by the image.

lto optimization cause that local symbol RTC_IRQHandler(from rtc_api.c)
somehow interferes with global symbol RTC_IRQHandler (from startup_stm32f070xb.S)

Changing local RTC_IRQHandler to _RTC_IRQHandler fixes problem

maciejbocianski committed:
ARMC6 keep __user_setup_stackheap symbol when LTO enabled

Fix for the error L6915E: Library reports error: Heap was used, but no heap region was defined

maciejbocianski committed:
add dummy SUPER_REALLOC/CALLOC calls to alloc wrappers

This code prevents the ARMC6 compiler/linker from removing
SUB_REALLOC/CALLOC symbols from image when LTO is enabled

Fixes below error:
L6137E: Symbol $Sub$$calloc was not preserved by the LTO codegen but is needed by the image.

kjbracey commented on platform/source/mbed_alloc_wrappers.cpp:
```diff
@@ -347,6 +347,16 @@ extern "C" void *SUB_REALLOC(void *ptr, size_t size)
+            // this code will never be executed
+            // it's just to tell the compiler/linker to preserve SUB_REALLOC symbol
+            // when LTO enabled
+            SUPER_REALLOC(NULL, 0);
```
Maybe I'm being dense, but why aren't these calls using `SUPER_REALLOC` and `SUPER_CALLOC` anyway? Why are they messing around re-implementing them?

kjbracey replied:
Okay, I see the point - if you did use them, you'd have to reimplement the stuff in your malloc wrapper. Probably a bit bigger overall.

kjbracey replied:
I'm not sure how that fix actually works, mind - so you've inserted a (sort-of) call to the $$Super, but how does that persuade the LTO to retain the $$Sub?

Or maybe that activates a magic heuristic in the compiler - if a function calls any $$Super, assume it's referenced? Hmm. But then why not treat any $$Sub as referenced?

Can you explain how you got to this?

maciejbocianski replied:
@jamesbeyond found that difference between `SUB_CALLOC/REALOC` and `SUB_MALLOC/FREE`

And then by the analogy I have added dummy `SUPER_CALLOC/REALOC` calls inside `SUB_CALLOC/REALOC`

maciejbocianski committed:
CC3220SF_LAUNCHXL: keep ulDebugHeader symbol in LTO builds

Add a "used" attribute to ulDebugHeader (placed in signature_section section)
to fix ARMC6 build with the "-flto" flag. (Error: L6236E: No section matches
selector - no section to be FIRST/LAST. )

This attribute, attached to a function, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
EV_COG_AD3029LZ: keep IVT_NAME/blank_checksum symbols in LTO builds

Add a "used" attribute to IVT_NAME/blank_checksum
to fix ARMC6 build with the "-flto" flag. (Error: L6236E: No section matches
selector - no section to be FIRST/LAST. )

This attribute, attached to a function, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
EV_COG_AD4050LZ: keep IVT_NAME/blank_checksum symbols in LTO builds

Add a "used" attribute to IVT_NAME/blank_checksum
to fix ARMC6 build with the "-flto" flag. (Error: L6236E: No section matches
selector - no section to be FIRST/LAST. )

This attribute, attached to a function, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
MIMXRT1050_EVt: keep hyperflash_config/image_vector_table symbols in LTO builds

Add a "used" attribute to hyperflash_config/image_vector_table to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST. )

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
MSP432_LAUNCHPAD: keep interruptVectors symbol in LTO builds

Add a "used" attribute to interruptVectors to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST. )

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_IOT_M263A: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST. )

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_IOT_M487: keep __vector_handlers symbols in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST. )

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_M252KG: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST. )

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_PFM_M453: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST.)

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_PFM_NANO130: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST.)

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NUMAKER_PFM_NUC472: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST.)

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
NU_PFM_M2351_NPSA: keep __vector_handlers symbol in LTO builds

Add a "used" attribute to __vector_handlers to fix ARMC6 build with
the "-flto" flag.
(Error: L6236E: No section matches selector - no section to be FIRST/LAST.)

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
component PSA: keep SVCHandler_main/tfm_pendsv_do_schedule symbol in LTO builds

Add a "used" attribute to SVCHandler_main/tfm_pendsv_do_schedule to fix ARMC6 build with
the "-flto" flag.

This attribute, attached to a function/variable, means that code must be emitted
for the function even if it appears that the function is not referenced.

maciejbocianski committed:
armc6: make lto an optional profile

0xc0170 reviewed (approved):
Is there any docs PR for this profile extension?

bulislaw reviewed (approved):
Looks good, do we have docs following this PR?
